### PR TITLE
awards: datastreams: add OpenAIRE Project HTTP Reader

### LINKED ISSUE #324: Awards reader not relying on file system for jobs system and differential OpenAIRE update
* Automatically download the latest OpenAIRE Graph dataset.
* Allow to choose between the full dataset or the differential one.
* Stream it to the next reader/transformer without relying on the file system.

### PR DESCRIPTION
**Pull request by ptamarit**
:heart: Thank you for your contribution!

### Description

* Fixes #324 
    > **Quoted description**:
    > * Automatically download the latest OpenAIRE Graph dataset.
    > * Allow to choose between the full dataset or the differential one.
    > * Stream it to the next reader/transformer without relying on the file system.
* How to review this Pull Request:
    * This Pull Request is separated in **4 logical commits**, I recommended to **review them one by one**.
* The **changes can be tested** in an InvenioRDM instance by
    * Integrating the changes from
        * inveniosoftware/invenio-app-rdm#2676
        * #328
    * doing:
        ```bash
        $ invenio-cli shell
        $ invenio vocabularies convert -v awards -o diff -t awards.yaml
        # OR
        $ invenio vocabularies convert -v awards -o full -t awards.yaml
        ```

### Checklist

Ticks in all boxes and 🟢 on all GitHub actions status checks are required to merge:

- [x] I'm aware of the [code of conduct](https://inveniordm.docs.cern.ch/contribute/code-of-conduct/).
- [x] I've created [logical separate commits](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commits) and followed the [commit message format](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commit-message).
- [x] I've added relevant test cases.
- [x] I've added relevant documentation.
- [x] I've marked [translation strings](https://inveniordm.docs.cern.ch/develop/howtos/i18n/).
- [x] I've identified the [copyright holder(s)](https://inveniordm.docs.cern.ch/contribute/copyright-policy/) and updated copyright headers for touched files (>15 lines contributions).
- [x] I've NOT included third-party code (copy/pasted source code or new dependencies).
    * If you have added [third-party code (copy/pasted or new dependencies)](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#third-party-codedependencies), please reach out to an [architect](https://github.com/orgs/inveniosoftware/teams/architects/members).

**Frontend**

- [x] I've followed the [CSS/JS](https://inveniordm.docs.cern.ch/develop/best-practices/css-js/) and [React](https://inveniordm.docs.cern.ch/develop/best-practices/react/) guidelines.
- [x] I've followed the [web accessibility](https://inveniordm.docs.cern.ch/develop/best-practices/accessibility/) guidelines.
- [x] I've followed the [user interface](https://inveniordm.docs.cern.ch/develop/best-practices/ui/) guidelines.

**Reminder**

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

### CODE REVIEW

File: invenio_vocabularies/contrib/awards/datastreams.py [hunk 1]
```diff
@@ -1,22 +1,79 @@
...
+        """Reads the latest OpenAIRE Graph Dataset project tar file from Zenodo and yields an in-memory binary stream of it."""
+        if item:
+            raise NotImplementedError(
+                "OpenAIREProjectHTTPReader does not support being chained after another reader"
+            )
+
+        if self._origin == "full":
+            # OpenAIRE Graph Dataset
+            api_url = "https://zenodo.org/api/records/3516917"
+        elif self._origin == "diff":
```
Comment (ptamarit) [inline invenio_vocabularies/contrib/awards/datastreams.py:41]: Not sure if the `"diff"` name is the best choice. Happy to hear your thoughts.

File: invenio_vocabularies/contrib/awards/datastreams.py [hunk 2]
```diff
@@ -1,22 +1,79 @@
...
+        )
+
+    def read(self, item=None, *args, **kwargs):
+        """Reads the latest OpenAIRE Graph Dataset project tar file from Zenodo and yields an in-memory binary stream of it."""
+        if item:
+            raise NotImplementedError(
+                "OpenAIREProjectHTTPReader does not support being chained after another reader"
+            )
+
+        if self._origin == "full":
```
Comment (ptamarit) [inline invenio_vocabularies/contrib/awards/datastreams.py:38]: FYI, [`OrcidHTTPReader` uses another approach, with `test_mode` and then setting `origin`](https://github.com/inveniosoftware/invenio-vocabularies/blob/master/invenio_vocabularies/contrib/names/datastreams.py#L25-L28), but I tested it and it did not work for me.

File: invenio_vocabularies/datastreams/readers.py [hunk 3]
```diff
@@ -80,7 +79,12 @@ def _iter(self, fp, *args, **kwargs):
     def read(self, item=None, *args, **kwargs):
         """Opens a tar archive or uses the given file pointer."""
         if item:
-            yield from self._iter(fp=item, *args, **kwargs)
+            if isinstance(item, tarfile.TarFile):
+                yield from self._iter(fp=item, *args, **kwargs)
+            else:
+                # If the item is not already a TarFile (e.g. if it is a BytesIO), try to create a TarFile from the item.
+                with tarfile.open(mode=self._mode, fileobj=item) as archive:
```
Comment (ptamarit) [inline invenio_vocabularies/datastreams/readers.py:86]: For `ZipFile`, the constructor could take filenames and `BytesIO` regardless.
Here, if we wanted to also support filenames, we would need to do an extra check and call `tarfile.open` differently (not using the `fileobj` parameter) like so:

```python
elif isinstance(item, os.PathLike):
    # If the item is not already a TarFile (e.g. if it is a PathLike), try to create a TarFile from the item.
    with tarfile.open(item, mode=self._mode) as archive:
        yield from self._iter(fp=archive, *args, **kwargs)
```

However, for now our only use case is getting a `BytesIO`, so I would say that we can keep it like this for now.

File: invenio_vocabularies/contrib/awards/datastreams.py (hunk 2 above)
Comment (slint) [reply invenio_vocabularies/contrib/awards/datastreams.py:38]: Yeah, the extra parameters are mainly purposed for when defining new configs. I think keeping it simple with reusing `origin` is probably the easiest for now.

**Review by slint — APPROVED** [review]

**Pull request merged**